services/src/main/java/org/keycloak/jose/jws/TokenSignature.java

```diff
@@ -0,0 +1,97 @@
+import org.keycloak.crypto.KeyUse;
+import org.keycloak.crypto.KeyWrapper;
+import org.keycloak.jose.jws.JWSSignatureProvider;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.provider.ProviderFactory;
+import org.keycloak.representations.JsonWebToken;
+import org.keycloak.representations.RefreshToken;
+import org.keycloak.util.TokenUtil;
+
+// KEYCLOAK-7560 Refactoring Token Signing and Verifying by Token Signature SPI
+
+public class TokenSignature {
+
+    private static final Logger logger = Logger.getLogger(TokenSignature.class);
+
+    KeycloakSession session;
+    RealmModel realm;
+    String sigAlgName;
+
+    public static TokenSignature getInstance(KeycloakSession session, RealmModel realm, String sigAlgName) {
+        return new TokenSignature(session, realm, sigAlgName);
+    }
+
+    public TokenSignature(KeycloakSession session, RealmModel realm, String sigAlgName) {
+        this.session = session;
+        this.realm = realm;
+        this.sigAlgName = sigAlgName;
+    }
+
+    public String sign(JsonWebToken jwt) {
+        TokenSignatureProvider tokenSignatureProvider = getTokenSignatureProvider(sigAlgName);
+        if (tokenSignatureProvider == null) return null;
+
+        KeyWrapper keyWrapper = session.keys().getActiveKey(realm, KeyUse.SIG, sigAlgName);
+        if (keyWrapper == null) return null;
+
+        String keyId = keyWrapper.getKid();
+        Key signKey = keyWrapper.getSignKey();
+        String encodedToken = new JWSBuilder().type("JWT").kid(keyId).jsonContent(jwt).sign((JWSSignatureProvider)tokenSignatureProvider, sigAlgName, signKey);
+        return encodedToken;
+    }
+
+    public boolean verify(JWSInput jws) throws JWSInputException {
+        TokenSignatureProvider tokenSignatureProvider = getTokenSignatureProvider(sigAlgName);
+        if (tokenSignatureProvider == null) return false;
+
+        KeyWrapper keyWrapper = null;
+        // Backwards compatibility. Old offline tokens didn't have KID in the header
+        if (jws.getHeader().getKeyId() == null && isOfflineToken(jws)) {
+            logger.debugf("KID is null in offline token. Using the realm active key to verify token signature.");
+            keyWrapper = session.keys().getActiveKey(realm, KeyUse.SIG, sigAlgName);
+        } else {
+            keyWrapper = session.keys().getKey(realm, jws.getHeader().getKeyId(), KeyUse.SIG, sigAlgName);
+        }
+        if (keyWrapper == null) return false;
+
+        return tokenSignatureProvider.verify(jws, keyWrapper.getVerifyKey());
+    }
+
+    private static final String COMPONENT_SIGNATURE_ALGORITHM_KEY = "org.keycloak.jose.jws.TokenSignatureProvider.algorithm";
+
+    @SuppressWarnings("rawtypes")
+    private TokenSignatureProvider getTokenSignatureProvider(String sigAlgName) {
+        List<ComponentModel> components = new LinkedList<>(realm.getComponents(realm.getId(), TokenSignatureProvider.class.getName()));
```

Comment:
`getTokenSignatureProvider` is called rather often and would benefit from caching `c.getProviderId()` object per `sigAlgName`.

Reply:
To make new provider instance by its corresponding provider factory, need to use its corresponding ComponentModel instance. Therefore, whole ComponentModel c needs to be cached instead of c.getProviderId().

Or how about caching the provider instance per sigAlgName?

Reply:
Yes. I agree with you. So I will not modify here and make a JIRA issue for its optimization.